Signal/src/environment/PushRegistrationManager.swift

```diff
@@ -0,0 +1,287 @@
+            return false
+        }
+
+        guard notificationSettings.types == [] else {
+            return false
+        }
+
+        return true
+    }
+
+    private func registerForVanillaPushToken() -> Promise<String> {
+        Logger.info("\(self.TAG) in \(#function)")
+        AssertIsOnMainThread()
+
+        guard self.vanillaTokenPromise == nil else {
+            let promise = vanillaTokenPromise!
+            assert(promise.isPending)
+            Logger.info("\(TAG) alreay pending promise for vanilla push token")
+            return promise.then { $0.hexEncodedString }
+        }
+
+        // No pending vanilla token yet. Create a new promise
+        let (promise, fulfill, reject) = Promise<Data>.pending()
+        self.vanillaTokenPromise = promise
+        self.fulfillVanillaTokenPromise = fulfill
+        self.rejectVanillaTokenPromise = reject
+        UIApplication.shared.registerForRemoteNotifications()
+
+        let kTimeout: TimeInterval = 10
+        let timeout: Promise<Data> = after(seconds: kTimeout).then { throw PushRegistrationError.timeout }
+        let promiseWithTimeout: Promise<Data> = race(promise, timeout)
+
+        return promiseWithTimeout.recover { error -> Promise<Data> in
+            switch error {
+            case PushRegistrationError.timeout:
+                if self.isSusceptibleToFailedPushRegistration {
+                    // If we've timed out on a device known to be susceptible to failures, quit trying
+                    // so the user doesn't remain indefinitely hung for no good reason.
+                    throw PushRegistrationError.pushNotSupported(description: "Device configuration disallows push notifications")
+                } else {
+                    // Sometimes registration can just take a while.
+                    // If we're not on a device known to be susceptible to push registration failure,
+                    // just return the original promise.
+                    return promise
+                }
+            default:
+                throw error
+            }
+        }.then { (pushTokenData: Data) -> String in
+            if self.isSusceptibleToFailedPushRegistration {
+                // Sentinal in case this bug is fixed.
+                owsFail("Device was unexpectedly able to complete push registration even though it was susceptible to failure.")
+            }
+
+            Logger.info("\(self.TAG) successfully registered for vanilla push notifications")
+            return pushTokenData.hexEncodedString
+        }.always {
+            self.vanillaTokenPromise = nil
+        }
+    }
+
+    private func registerForVoipPushToken() -> Promise<String> {
+        AssertIsOnMainThread()
+        Logger.info("\(self.TAG) in \(#function)")
+
+        guard self.voipTokenPromise == nil else {
+            let promise = self.voipTokenPromise!
+            assert(promise.isPending)
+            return promise.then { $0.hexEncodedString }
```

Comment:
I'm not a big fan of `$0`.  More readable to name the variable.

Reply:
Agreed that `$0` *can* be damaging for readability, but for *really* short blocks, (like 1 or 2 lines), I think it can actually help readability when used judiciously.

Similarly, depending on the context, sometimes you want something like:

```
let myFoo = Foo()
let myBar = myFoo.bar
let myQux = myBar.qux
return myQux
```

And sometimes you want something more like:

```
return Foo().bar.qux
```